src/portalbot_service.py

```diff
@@ -0,0 +1,513 @@
+        message_type = message.get("type")
+        data = message.get("data", {})
+
+        logger.info(f"Received from public server: {message_type}")
+
+        if message_type == "connected":
+            # Server acknowledged connection, now identify as robot and join space
+            await self.identify_as_robot()
+
+        elif message_type == "joined_space":
+            logger.info(f"Successfully joined space: {data.get('space')}")
+
+        elif message_type == "control_request":
+            # Someone wants to control the robot
+            await self.handle_control_request(data)
+
+        elif message_type == "control_released":
+            # Controller released control
+            await self.handle_control_released(data)
+
+        elif message_type == "remote_command":
+            # Remote control command to relay to central_hub
+            await self.handle_remote_command(data)
+
+        elif message_type == "offer":
+            # WebRTC offer from remote operator
+            # TODO: Implement WebRTC handling
+            logger.info("Received WebRTC offer (WebRTC not yet implemented)")
+
+        elif message_type == "ice_candidate":
+            # WebRTC ICE candidate
+            # TODO: Implement WebRTC handling
+            logger.debug("Received ICE candidate (WebRTC not yet implemented)")
+
+        elif message_type == "error":
+            logger.error(f"Error from public server: {data.get('message')}")
+
+    async def identify_as_robot(self):
+        """Send robot identification to public server"""
+        await self.send_to_public_server(
+            "robot_identify",
+            {
+                "robot_id": self.config.robot_id,
+                "robot_name": self.config.robot_name,
+                "space": self.config.space_id,
+                "secret_key": self.secret_key,
+            },
+        )
+
+    async def handle_control_request(self, data: dict):
+        """Handle a control request from a remote operator"""
+        controller_id = data.get("controller_id")
+        logger.info(f"Control requested by: {controller_id}")
+
+        self.control_state = ControlState.CONTROL_REQUESTED
+        self.controller_id = controller_id
+
+        # TODO: Start validation process (audio + face detection)
+        # For now, automatically grant control
+        await asyncio.sleep(0.5)  # Simulate validation
+
+        # Grant control
+        self.control_state = ControlState.CONTROLLED
+        await self.send_to_public_server(
+            "control_granted", {"controller_id": controller_id}
+        )
+        logger.info(f"Control granted to: {controller_id}")
+
+    async def handle_control_released(self, data: dict):
+        """Handle control being released"""
+        logger.info("Control released")
+        self.control_state = ControlState.IDLE
+        self.controller_id = None
+
+    async def handle_remote_command(self, data: dict):
```

Comment:
we need to limit this to just accepting `servo_angles` that get relayed to central_hub.